cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -122,15 +161,23 @@ public void setAuditPublisher(AuditPublisher auditPublisher) {
    * @throws NotFoundException if either the namespace or dataset instance is not found,
    * @throws IOException if there is a problem in making an HTTP request to check if the namespace exists.
    */
-  public DatasetMeta get(Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
+  public DatasetMeta get(final Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
+    // Application Deployment first makes a call to the dataset service to check if the instance already exists with
+    // a different type. To make sure that that call responds with the right exceptions if necessary, first fetch the
+    // meta from the cache and throw appropriate exceptions if necessary.
+    DatasetMeta datasetMeta;
     try {
-      return metaCache.get(instance);
+      datasetMeta = metaCache.get(instance);
     } catch (ExecutionException e) {
-      if ((e.getCause() instanceof Exception) && (e.getCause() instanceof HttpErrorStatusProvider)) {
-         throw (Exception) e.getCause();
+      Throwable cause = e.getCause();
+      if ((cause instanceof Exception) && (cause instanceof HttpErrorStatusProvider)) {
+         throw (Exception) cause;
       }
       throw e;
     }
+    // Only return the above datasetMeta if authorization succeeds
+    ensureAccess(instance.toEntityId());
```

Comment:
I think the scenario that the existing code in the deployment pipeline is: Alice deploys App1, which creates dataset D1. Someone (Alice or otherwise) tries to deploy App2, which contains D1, but with a different, incompatible type. The code we have today in the app deployment pipeline checks for this behavior, and there are tests to make sure that we throw a NotFoundException/404. 

If we add an authorization check in the beginning, we may be checking the privileges of a non-existing entity, and that operation will always result in an UnauthorizedException. When you're deploying datasets for the first time, this will always be the case, since the entity does not exist.